v2/internal/util/interval/interval_calculator.go

```diff
@@ -0,0 +1,170 @@
+	if hasRequeueDelayOverride && isRequeueing {
+		result.RequeueAfter = i.requeueDelayOverride
+		result.Requeue = true
+	}
+	return result, nil
+}
+
+func (i *calculator) failureResult(req ctrl.Request, err error) (ctrl.Result, error) {
+	exp := i.failures[req]
+	i.failures[req] = i.failures[req] + 1
+
+	readyErr, ok := conditions.AsReadyConditionImpactingError(err)
+	if !ok {
+		// NotFound is a superfluous error as per https://github.com/kubernetes-sigs/controller-runtime/issues/377
+		// The correct handling is just to ignore it and we will get an event shortly with the updated version to patch
+		// We must also ignore conflict here because updating a resource that
+		// doesn't exist returns conflict unfortunately: https://github.com/kubernetes/kubernetes/issues/89985. This is OK
+		// to ignore because a conflict means either the resource has been deleted (in which case there's nothing to do) or
+		// it has been updated, in which case there's going to be a new event triggered for it and we can count this
+		// round of reconciliation as a success and wait for the next event.
+		return ctrl.Result{}, kubeclient.IgnoreNotFoundAndConflict(err)
+	}
+
+	// Now we have a readyErr
+	if readyErr.Severity == conditions.ConditionSeverityError {
+		// Severity error is fatal, return fast and block requeue
+		// Since this is fatal, stop tracking the req
+		delete(i.failures, req)
+		return ctrl.Result{}, nil
+	} else if readyErr.Severity == conditions.ConditionSeverityWarning {
+		switch readyErr.RetryClassification {
+		case conditions.RetrySlow:
+			delay := i.calculateExponentialDelay(i.baseDelay, exp, i.maxSlowDelay)
+			return ctrl.Result{RequeueAfter: delay}, nil
+		case conditions.RetryFast:
+			delay := i.calculateExponentialDelay(i.baseDelay, exp, i.maxFastDelay)
+			return ctrl.Result{RequeueAfter: delay}, nil
+		case conditions.RetryNone:
+			// This shouldn't happen, return an error
+			return ctrl.Result{}, errors.New("didn't expect RetryNone classification for error")
+		default:
+			// This shouldn't happen, return an error
+			return ctrl.Result{}, errors.Errorf("unknown RetryClassification %q", readyErr.RetryClassification)
+		}
+	}
+
+	// This shouldn't happen, return an error
+	return ctrl.Result{}, errors.Errorf("Error with severity %q is unexpected", readyErr.Severity)
+}
+
+func (i *calculator) makeSuccessResult() ctrl.Result {
+	result := ctrl.Result{}
+	// This has a RequeueAfter because we want to force a re-sync at some point in the future in order to catch
+	// potential drift from the state in Azure. Note that we cannot use mgr.Options.SyncPeriod for this because we filter
+	// our events by predicate.GenerationChangedPredicate and the generation will not have changed.
+	if i.syncPeriod != nil {
+		result.RequeueAfter = randextensions.Jitter(i.rand, *i.syncPeriod, 0.1)
+	}
+
+	return result
+}
+
+func (i *calculator) calculateExponentialDelay(base time.Duration, exp int, max time.Duration) time.Duration {
```

Comment:
Shouldn't this have some jitter too?

Reply:
Maybe, but the existing interface which we're replacing (rate_limiter) didn't, so we're in no different a situation jitter-wise now than we were before.